Me: Allow Gravatar uploads

sirbrillig opened the pull request:
There is already a Gravatar REST API endpoint to upload images, but it requires authentication with an OAuth token. Clients like Calypso use the WPCOM REST API with a Cookie instead of a token, and therefore they cannot use the Gravatar API directly.

In D6120, we added a WPCOM endpoint that effectively proxies the request to the Gravatar API by providing the OAuth token itself.

This PR changes the `EditGravatar` component to use this new endpoint and enables the component in development.

## Testing

Visit http://calypso.localhost:3000/me and you should see Gravatar section in the Profile box:

<img width="184" alt="screen shot 2017-06-28 at 4 58 42 pm" src="https://user-images.githubusercontent.com/2036909/27659914-24792c6e-5c23-11e7-9ff5-549142850302.png">

Clicking on the image there, you should be able to follow the flow to upload a new image. When the upload is complete you should see a completion message:

<img width="561" alt="screen shot 2017-06-28 at 5 09 25 pm" src="https://user-images.githubusercontent.com/2036909/27660261-9372cf52-5c24-11e7-8ee7-026aa662fa0d.png">

Also view the Gravatar page at https://gravatar.com and be sure that the uploaded image is there:

<img width="448" alt="screen shot 2017-06-28 at 5 12 47 pm" src="https://user-images.githubusercontent.com/2036909/27660407-0cde9d12-5c25-11e7-864a-c925ed307283.png">

v18 commented on client/state/current-user/gravatar-status/actions.js:
```diff
@@ -20,20 +20,13 @@ import {
-			.post( 'https://api.gravatar.com/v1/upload-image' )
-			.send( data )
-			.set( 'Authorization', 'Bearer ' + bearerToken )
+		return wpcom.undocumented().uploadGravatar( email, file )
```
Should we match the argument order (`file`, `email`) in wpcom.undocumented().uploadGravatar?

v18 replied:
I think we are trying to move away from using `wpcom` and use the [data layer](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer) instead. Maybe this is a good time to switch `uploadGravatar` over? cc @gwwar and @dmsnell for advice on this.

sirbrillig replied:
That's probably a good idea. 👍

sirbrillig replied:
I'd be happy to do so, but I've never used it. I suppose it depends on how quickly we want this to work. If there's no rush, it would be awesome to remove this thunk altogether and use the data layer to handle the network requests. On the other hand, since it has no effect on the user experience that could be done in a separate PR. I'll look into what using the new system would look like.

v18 commented on config/development.json:
```diff
@@ -95,7 +95,7 @@
 		"manage/themes/logged-out": true,
 		"manage/themes/upload": true,
 		"me/account": true,
-		"me/edit-gravatar": false,
```
We should make sure to get the API changes in before merging this.

sirbrillig replied:
API changes are deployed.

v18 commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -713,6 +713,26 @@ Undocumented.prototype.metaKeyring = function( fn ) {
+ * @return {Promise} A Promise to resolve when complete
+ * @api public
+ */
+Undocumented.prototype.uploadGravatar = function( email, file, fn ) {
```
Why do we need the `fn` parameter here?

sirbrillig replied:
It's just the pattern in all the functions in `wpcom`. If a `fn` callback is provided, it will use the callback; otherwise it will return a Promise. tbh, I'd rather just skip the callback and rely on the promise instead.

v18 replied:
Yeh, I agree with that - no reason to introduce callbacks here.

v18 commented:
Ran into an issue when trying to upload an image while offline - it shows the upload was successfully when the network request in fact fails. Steps to repro:
1. Select a new image to upload
2. Before clicking 'Change my photo', turn off the network in chrome dev tools
3. Click 'Change my photo'
4. See the success notice saying the new Gravatar was uploaded
5. See that the network request in fact failed, and no new Gravatar was uploaded to Gravatar.com

sirbrillig commented:
I modified the PR to use the data-layer instead of adding a method to `Undocumented`.

This exposed a bug (I think) in the data-layer's `http` helper method; `apiNamespace` is not correctly applied to requests. I added a fix for that in fb81fc997c0ac92fd43b1358210da36d49f9d05e but that should probably be investigated and made into its own PR (cc @dmsnell).

I haven't yet investigated the offline network status. The handling might already be taken care of by the data-layer, but I'll need to do some tests.

sirbrillig commented:
Thank you for the thoughts, @dmsnell! 

I reviewed the bug I found and did some more research. I now know why it's happening and I was able to work around it without modifying the data-layer. Here's the summary:

In wpcom.js, the `sendRequest()` function [copies properties like `apiNamespace`](https://github.com/Automattic/wpcom.js/blob/0902b5eb8fcc4a3dd6760bb7a315e77315ead0ad/lib/util/send-request.js#L53) from `query` to `params`. This allows those params to be set in the `query` instead. (I'm not sure what the intent of this was, but it's how the code works.)

However, the function `post()`, which calls `sendRequest()`, first changes some of its arguments. In particular, [if there is no callback and no `body`](https://github.com/Automattic/wpcom.js/blob/ab065dfc1e515aba6e36cb9b6b438366dff9bc3c/lib/util/request.js#L43), it will remove `query` entirely. This change will prevent the copying of `apiNamespace` above.

The result is that if `apiNamespace` is set directly in `params`, the API path will correctly use it, but if it is set in `query`, the path will be incorrect unless `body` or a callback are also set.

To work around this, I simply set an empty `body` when calling `http()`. This is not immediately obvious, but the root cause is the awkward behaviors of wpcom.js ([already](https://github.com/Automattic/wpcom.js/pull/210) in [discussion](https://github.com/Automattic/wpcom.js/pull/211)) and I'm not sure _if_ or _how_ it should be hidden in Calypso.

> Although the docs may still be misleading we have removed the next() call from the data layer API. We will want to remove those calls and parameters. In uploadFailed we should use store.dispatch() instead.

That's good to know! I can see how `next` could be confusing, since its behavior _seems_ to be the opposite of `next` in typical middleware, skipping the other handlers rather than moving to the next one (it's not really opposite because all of the data-layer is one middleware, but each handler of the data-layer itself feels like middleware).

That said, if I remove using `next`, how could I prevent an infinite loop when handling an action that I also want to be handled in the reducers? Is the intent not to have actions that are handled in both places?

> Let's see if we can focus on naming the handlers in the data layer in ways which are meaningful to the application. uploadGravatar is great! What does uploadComplete mean though?

This is a very interesting thought! `announceFailure` says what the action is going to do, while `uploadFailed` is more of a state result. I could argue that an action like `uploadFailed` isn't intended to do anything, rather that reducers should take it as information and then decide what to do, but perhaps that's not the role of reducers. If you have time I'd love to hear more about how you see this concept. In the mean time, I've changed these handlers to `announceSuccess` and `announceFailure`.

> Are we reusing existing reducers? The data layer is designed to eliminate the creation of _RECEIVE, _SUCCESS, and _FAILURE action types. Instead of issuing these (which carry no real semantic meaning) we can instead focus on meaningful action types for the app, such as USER_SET_GRAVATAR or use our global notices to announce failures.

I think I see where you're going here. I'm inclined to avoid changing the way success/failure is reported in this PR just to keep it at a reasonable size. Changing that makes a lot of sense, but would require not only a change to the reducer, but also to the state tree, selectors, and the components which rely on that state. I'd suggest a follow-up to this PR which tackles that refactor.

dmsnell commented on client/state/current-user/gravatar-status/test/actions.js:
```diff
@@ -42,22 +42,22 @@ describe( 'actions', () => {
-					.post( '/v1/upload-image' )
+					.post( '/wpcom/v2/gravatar-upload' )
 					.reply( 200, 'Successful request' );
 			} );
```
note: with the use of `http()` in the data layer we should be able to completely eliminate the use of `nock()` - just make sure that the dispatched action has the right path and properties, etc…

dmsnell commented on client/state/data-layer/wpcom/gravatar-upload/index.js:
```diff
@@ -0,0 +1,62 @@
+		],
+	}, action ) );
+
+	next( action );
```
without this `next()` available the original `_REQUEST` action will automatically pass through to the reducers. if we have to dispatch different actions (for example, suppose we had to also grab user details and dispatch `USER_REQUEST`) then we just "start over" with `store.dispatch`.

on the other hand, if we need to pass through an action which could cause a cycle/loop then we must wrap it in `local` so that it bypasses the data layer

```js
store.dispatch( local( action ) )
```

only in rare cases should we need to send off a local action

sirbrillig replied:
Ah, good to know! I thought that if an action was "caught" by middleware then it was removed from the dispatch chain and would not make it to the reducers unless explicitly dispatched again.

sirbrillig commented:
> Ran into an issue when trying to upload an image while offline - it shows the upload was successful when the network request in fact fails. 

This is still an issue. Researching what can be done...

dmsnell commented:
> data-layer apparently treats a failed network connection as a success,

this is wrong. or at least, if it is it's a bug. a request which fails to make it out should never come back as a success. in fact, the data layer mirrors `wpcom.js` (which is dubious for sure): if `wpcom.js` returns an error then the data layer should call a failure; if it doesn't return an error then we should call a success.

@samouri has been exploring a validator which will also cause a non-error `wpcom.js` request to fail if it doesn't validate. for example, some of our APIs return 2xx responses which contain nothing but an error.

what is leading you to believe that it's treating a failed network connection as a success?

sirbrillig commented:
> what is leading you to believe that it's treating a failed network connection as a success?

I'd like to be mistaken, for sure! Interestingly, I tried setting up a minimal test with wpcom.js which implies that it does not have this behavior, so I'm not sure where it's coming from. Nevertheless, try the following:

Using this PR...

1. Visit http://calypso.localhost:3000/me
1. Click the Gravatar image in the main page section.
1. Select a new image to upload
2. Before clicking 'Change my photo', turn off the network in chrome dev tools
3. Click 'Change my photo'
4. See the success notice saying the new Gravatar was uploaded, even though it was not.

I added some debugging and as @v18 said [above](https://github.com/Automattic/wp-calypso/pull/15636#issuecomment-312273094), the [handler](https://github.com/Automattic/wp-calypso/blob/a70587157b33783ba4d5e0e4a1e136680d459d51/client/state/data-layer/wpcom-http/index.js#L57) reports `error` as `null` and `data` as `{}` even though there was no network connection.

dmsnell commented:
The problem is that `wpcom-proxy-request` [treats unsent messages as successes](https://github.com/Automattic/wpcom-proxy-request/blob/912fdb1fa08df0a4e24cf944cd70119bbe0618ba/index.js#L473)

In your upload request the underlying XHR object has a `readyState` value of `0`, which [indicates that the request has not been sent](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState) and yet `wpcom-proxy-request` things that because there's no `status` code available it must be a success (yes, it's using type-coercing equality making this also a type-related bug that could have/should have been prevented with static typing).

In my assessment we need to fix `wpcom-proxy-request` because the data layer will not be able to know when it is misleading or accurately reporting. I will try and get a PR started in that library though it may take some time because I'm not familiar with it.

Thanks for finding this bug!

v18 commented:
@dmsnell From what I can see the offline bug is coming from https://github.com/Automattic/wpcom.js/blob/master/lib/util/send-request.js#L91

We look for an error in the request, and if it's not found we resolve. In this case, err = null and res = {}.

> In my assessment we need to fix `wpcom-proxy-request` because the data layer will not be able to know when it is misleading or accurately reporting.

Are we using `wpcom-proxy-request` in the data layer - my debugging led me directly to wpcom.js.

dmsnell commented:
> Are we using wpcom-proxy-request in the data layer - my debugging led me directly to wpcom.js.

`wpcom.js` gets wrapped by `wpcom-proxy-request` and it hides lots of information from the actual XHR. in Calypso we get a fake XHR object and the error messages conveying this failed state get lost in the translation. I have submitted automattic/wpcom-proxy-request#20 in an attempt to fix it but I don't know what other things that "fix" could break.

the heart of the problem as best I can tell is that we are assuming that requests are successful unless they carry an explicitly fail-state status code. of course, when requests just die because the network connection has dropped then we get no status code (or, more technically it appears that through the mess we are getting a status code of `0`, which coerced to `null` in the weak-equality check being used).

Had we the actual XHR object which made the request this should not be a problem. It's because the XHR is faked (incompletely so) that we lose the information and fail.

v18 commented:
@sirbrillig @dmsnell do you think we should wait for the offline fix, or work around it for now by checking for the success property? If the fix might take a while to land, I think it's best to move forward with the workaround.

dmsnell commented on client/state/data-layer/wpcom/gravatar-upload/index.js:
```diff
@@ -0,0 +1,60 @@
+}
+
+function announceSuccess( { dispatch }, { file } ) {
+	const fileReader = new FileReader( file );
```
this is IE10+ only, is that okay? should we be guarding against it? I thought we supported IE9 but I can't remember

v18 replied:
We support only IE11+. On IE 10, Calypso shows the `Unsupported Browser` screen.

dmsnell commented:
@sirbrillig I didn't mean to suggest that you should kill all of your tests 😆 

you can still verify that the actions trigger the appropriate network calls - you just don't have to actually mock the network since it's all just data

sirbrillig commented:
I had to rewrite the tests because really what we're testing now is the data-layer part of the code, rather than the triggering action.

@v18 I added a work-around in the code for now.

IMO this feature is ready. Let me know what you think.

v18 commented:
Looking great so far. A few notes:
- I noticed that I had previously misspelled `receive` in `calypso_edit_gravatar_file_recieve_failure` -- do you think it's ok to fix in this PR? Happy to do it in a separate PR later, too.
- `unsuccessful_http_response` now refers to both non-2xx responses, and also some 200 responses. Should we change it to `unsuccessful_response`?
- I noticed that the event `calypso_edit_gravatar_upload_start` is emitted many times during a successful upload, instead of once. It's problematic since it will bloat the upload start stats. This happens because the `onProgress` event causes the `action` object to be dispatched ([ref](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-http/index.js#L78)) (related to #14538).  I don't see an easy workaround for this, what do you think @dmsnell ? 

Progress event n:
<img width="994" alt="progress-1" src="https://user-images.githubusercontent.com/11487924/27846300-1aeeb32c-6104-11e7-913d-d871b4516144.png">
Progress event n+1:
<img width="1019" alt="progress-2" src="https://user-images.githubusercontent.com/11487924/27846301-1af22232-6104-11e7-9cc7-bc62e85567ea.png">

dmsnell commented:
>  I don't see an easy workaround for this, what do you think @dmsnell ?

Easy workaround in [#15730](https://github.com/Automattic/wp-calypso/pull/15730/files#diff-2bd08d4c813d9f79ed9ef35b675e71eeR29) :smile:

dmsnell commented on client/state/data-layer/wpcom/gravatar-upload/index.js:
```diff
@@ -0,0 +1,74 @@
+}
+
+export function announceSuccess( { dispatch }, { file }, next, data ) {
+	// Return a Promise for testing purposes
```
I understand the desire to test everything but this is starting to feel a little much just to be able to test. I'll try and vocalize a few observations:
  - We're already mocking the `FileReader` in the test. We're essentially mocking it here to with the awkward use of the Promise.
  - Handlers are meant to be treated fully asynchronously, so returning a value is definitely going to be adding examples to the code base that will confuse other people reading it.
  - The existing test in the data layer is an integration test, not a unit test - it's testing _both_ the behavior of the `FileReader` _and_ the behavior of the handler.
  - Maybe if we're going to mock it we should full-out mock it in the test which could leave this code in its earlier and simpler form.
  - Could we mock it in such a way that we could manually trigger `load` in the test? `middleware()()(); mockReader.finishLoading(); expect( dispatch ).to…`

Anyway, I'm just wanting to make sure we don't do something we really don't want to do because we think that we have to. In the end, I'd rather have no test than a confusing and fragile one.

sirbrillig replied:
Yeah. A few hours away from this code made me realize that I was already turning the async `FileReader` into a sync process in the tests. I didn't feel good about the Promise, but yesterday I couldn't see a better way. Thanks for pushing back on that.

dmsnell replied:
:+1: thanks for being so diligent on it all!

dmsnell commented on client/state/data-layer/wpcom/gravatar-upload/index.js:
```diff
@@ -0,0 +1,74 @@
+	return new Promise( ( resolve, reject ) => {
+		// Sometimes wpcom-proxy-request reports a failed connection as a success, so
+		// we test for the response here. See
+		// https://github.com/Automattic/wp-calypso/pull/15636#issuecomment-312860944
```
Would be great if we could just fix `wpcom-proxy-request` before this PR ships so that we don't need this workaround and associated tests.

dmsnell commented on client/state/data-layer/wpcom/gravatar-upload/index.js:
```diff
@@ -0,0 +1,74 @@
+		const fileReader = new FileReader( file );
+		fileReader.addEventListener( 'load', () => {
+			resolve( fileReader.result );
+		} );
```
Fun fact! There is no need for the curlies here. Since arrow-functions (in the absence of the curlies) return the evaluation of expression we can just provide the expression (even if we ignore the return value)

```js
fileReader.addEventListener( 'load', () => resolve( fileReader.result ) );
```

of course, this is just a fun fact. if we can get by without the promise it won't apply here specifically anymore

dmsnell commented on client/state/data-layer/wpcom/gravatar-upload/test/index.js:
```diff
@@ -0,0 +1,119 @@
+
+		expect( dispatch ).to.have.been.calledOnce;
+		expect( dispatch ).to.have.been.calledWith( sinon.match( { type: GRAVATAR_UPLOAD_REQUEST_FAILURE } ) );
+	} );
```
if we fix `wpcom-proxy-request` we won't need this test

dmsnell commented:
There are sill a few things I would love to work on before merging this. Please feel free to ping me in Slack if this is too _async_. I don't want to hold you up, but particularly with `wpcom-proxy-request` we've identified a bug that needs to be fixed and just working around it is going to do us harm in the long run when we do fix it.

sirbrillig commented:
I took away the Promise hacks, fixed the "receive" typo, and rebased to hopefully fix the multiple event issue you saw, @v18. 

I think that it would be nice to deal with the `wpcom-proxy-request` bug so we can remove that check from `announceSuccess` (and the associated test). That said, I'll be gone for all of next week, so I'll leave it to you to decide, @v18.

v18 commented:
> Easy workaround in #15730

@dmsnell even after using the fixes from #15730, I see the `calypso_edit_gravatar_upload_start` tracking event being fired multiple times:

<img width="770" alt="screen shot 2017-07-05 at 3 03 52 pm" src="https://user-images.githubusercontent.com/11487924/27880548-5d47bacc-6193-11e7-8dc9-c1514f0a8e87.png">

I haven't dug down to the root cause, but there are two (maybe separate) issues at play that I can see so far:

1. The event handler for progress events still gets attached in https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-http/index.js#L78. So each time there's a progress event, the handler still gets executed and dispatches an action with the analytics event. That feels to me like a bug. What do you think @dmsnell ?
2. Even if the event handler is not attached to progress events, an extra analytics event (`calypso_edit_gravatar_upload_start` in this case) is still fired - this is what you can see in the screenshot above. This also seems like a bug.

I don't think we can ship until only one `calypso_edit_gravatar_upload_start` event is fired per image upload.

gwwar commented:
@v18 oh nice find. Do you have a screenshot of the how many `GRAVATAR_UPLOAD_REQUEST` make it to the redux devtools? We might have a bug here with middleware ordering or with analytics middleware.

v18 commented on client/blocks/edit-gravatar/test/index.jsx:
```diff
@@ -42,13 +42,14 @@ describe( 'EditGravatar', function() {
 	} );
 
-	before( function() {
+	before( function( done ) {
```
I still occasionally get the timeout error:

`Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.`

It may be the case that we're sometimes running into an error when running `before()`, but it's hard to reproduce, so haven't caught it yet with a try-catch block.

sirbrillig replied:
After some digging I think that loading the modules inside this `before()` are just taking a long time to `require()` for some reason, which intermittently causes the tests to fail. I might just remove the `done` hack and create a separate issue for this.

dmsnell replied:
would be happy to discuss this. tests should never be intermittent or nondeterministic. also `done` is there for a very specific purpose (async flows) so it's not a hack to use it. I'm not sure how you were using it but you may be too hard on yourself :smile:

v18 commented:
@gwwar only one `GRAVATAR_UPLOAD_REQUEST` makes it to the redux dev tools:

<img width="461" alt="screen shot 2017-07-05 at 3 59 47 pm" src="https://user-images.githubusercontent.com/11487924/27882624-611a6d72-619b-11e7-99a4-bcfe6032dca4.png">

Before #15730, there were many `GRAVATAR_UPLOAD_REQUEST` being dispatched, so the fix there worked 👍 

Here's another screenshot with all the redux actions logged together with the analytics events (after removing the progress event handler for clarity):

<img width="804" alt="screen shot 2017-07-05 at 4 07 02 pm" src="https://user-images.githubusercontent.com/11487924/27882788-149c4802-619c-11e7-8122-4949be858cf5.png">

gwwar commented:
Current ordering looks like:
```jsx
const middlewares = [
		thunkMiddleware,
		noticesMiddleware,
		isBrowser && require( './happychat/middleware.js' ).default(),
		isBrowser && require( './analytics/middleware.js' ).analyticsMiddleware,
		require( './data-layer/wpcom-api-middleware.js' ).default,
		isBrowser && require( './v18lib/middleware.js' ).default,
		isBrowser && config.isEnabled( 'restore-last-location' ) && require( './routing/middleware.js' ).default,
		isBrowser && require( './data-layer/extensions-middleware.js' ).default,
		isAudioSupported && require( './audio/middleware.js' ).default,
		isBrowser && config.isEnabled( 'automated-transfer' ) && require( './automated-transfer/middleware.js' ).default,
	].filter( Boolean );
```

@v18 do you mind experimenting with moving analytics after the data-layer in `client/state/index` I'm thinking we might need to move the data-layer to right after the thunkMiddleware.

v18 commented:
@gwwar 👌! That fixes both problems with this new middleware order:

```
const middlewares = [
  thunkMiddleware,
  require( './data-layer/wpcom-api-middleware.js' ).default,
  noticesMiddleware,
  isBrowser && require( './happychat/middleware.js' ).default(),
  isBrowser && require( './analytics/middleware.js' ).analyticsMiddleware,
  isBrowser && require( './lib/middleware.js' ).default,
  isBrowser && config.isEnabled( 'restore-last-location' ) && require( './routing/middleware.js' ).default,
  isBrowser && require( './data-layer/extensions-middleware.js' ).default,
  isAudioSupported && require( './audio/middleware.js' ).default,
  isBrowser && config.isEnabled( 'automated-transfer' ) && require( './automated-transfer/middleware.js' ).default,
].filter( Boolean );
```

I imagine we also want the extensions data layer to come right after the wpcom data layer?

v18 commented:
@sirbrillig btw I can open a separate PR for the middleware changes.

dmsnell commented on client/state/current-user/gravatar-status/reducer.js:
```diff
@@ -16,11 +16,7 @@ export const isUploading = createReducer( false, {
-		};
-	}
+	[ GRAVATAR_UPLOAD_RECEIVE ]: ( state, action ) => ( { src: action.src } ),
 } );
```
do we really need `createReducer()` here? if all we're doing is this simple little thing, how about making it a simple little function?

```js
export const tempImage = ( state = null, { type, src } ) =>
	type === GRAVATAR_UPLOAD_RECEIVE
		? src
		: state;
```

Also, why are we storing this string inside an object instead of just storing the string? Seems like a bit hasty future-planning? If we _do_ predict needing more properties later on, that's what `combineReducers()` is for, so we can write our reducers as simple functions on simple data types and then compose them together.

dmsnell commented:
@sirbrillig it would be nice if you could look over the reducer comment I left. I don't think it warrants `createReducer()` - curious if you are willing to share: for what reason did you reach out to `createReducer()` in the first place?

All things considered this looks fine. The middleware change is good as I think the data layer does need to come first. It would be nice to wait to merge until `wpcom-proxy-request` is fixed and then you can skip testing for its bug but I won't stand in your way if you can't wait. Please though promise to remove the code if we get it fixed. Adding tests in one project for the bugs of another project is just stacking up maintenance burden :smile:

sirbrillig commented:
> @sirbrillig it would be nice if you could look over the reducer comment I left. I don't think it warrants createReducer() - curious if you are willing to share: for what reason did you reach out to createReducer() in the first place?

tbh, I've never used `createReducer`. This code was already in place. I was just simplifying it a bit since I was in the neighborhood. But maybe it's better if I leave that reducer alone and tackle it in a separate PR (eg: refactoring the upload process to be optimistic).

My plan is to wait until the `wpcom-proxy-request` work is complete before merging this (I'll see if I can help with that), since we also have to wait on the endpoint review itself and the middleware ordering change.

v18 commented on client/state/data-layer/wpcom/gravatar-upload/test/index.js:
```diff
@@ -0,0 +1,115 @@
+		const dispatch = spy();
+
+		announceSuccess( { dispatch }, action, noop, { success: true } );
+		expect( dispatch ).to.have.been.calledWith( sinon.match( { type: GRAVATAR_UPLOAD_RECEIVE, src: 'tempImageSrc' } ) );
```
nitpick: here I think you meant to use the variable `tempImageSrc` and not the string `'tempImageSrc'`.

sirbrillig replied:
Hm. I had put the string intentionally, but I suppose it could be either.

v18 replied:
All good! :)

sirbrillig commented:
Also worth discussing: this PR only enables the feature in development. Once all the dependencies are merged, do we want to enable it in production/desktop/etc. also in this PR or make that a separate PR? (I'm inclined to put it all here.)

v18 commented:
@sirbrillig my preference would be to get this PR merged, and handle enabling it in production in a separate PR. Some reasons:
1. Enabling this PR in production also involves removing some sections of code, which would need a separate 👀 , so would be good to do a clean small PR for that.
2. I found a few bugs in the feature (missing translation, clickable button when it should not be etc), that would be nice to get fixed before production.
3. Lower likelihood of this getting out of sync with master.
4. Easier to test the remaining blockers if this is in master.

sirbrillig commented:
Sounds good to me. I'd like to get D6120 (the endpoint) deployed first, then I'll remove the `success` hack and merge this.

sirbrillig commented:
The endpoint is deployed, and I've removed the `success` hack (saved [here](https://gist.github.com/sirbrillig/b1231c84c98d91397ffb5b2248b06d8e) in case we want it). I'm going to merge this shortly unless there are any objections, making it available in development only for further testing, etc.
